### PR TITLE
feat(expr): add support for make_date/time/timestamp

### PR DESCRIPTION
**Pull request by KeXiangWang**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

As title
https://www.postgresql.org/docs/current/functions-datetime.html

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [x] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [x] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

```
make_date ( year int, month int, day int ) → date
Create date from year, month and day fields (negative years signify BC)
make_date(2024, 1, 31) → 2024-01-31

make_time ( hour int, min int, sec double precision ) → time
Create time from hour, minute and seconds fields
make_time(1, 45, 30.2) → 01:45:30.2

make_timestamp ( year int, month int, day int, hour int, min int, sec double precision ) → timestamp
Create timestamp from year, month, day, hour, minute and seconds fields (negative years signify BC)
make_timestamp(2024, 1, 31, 1, 45, 30.2) → 2024-01-31 01:45:30.2
```

### CODE REVIEW

**Review by wangrunji0408 — APPROVED** [review]
Generally LGTM.

@xiangjinwu please double check. 🥹

File: src/expr/impl/src/scalar/make_time.rs [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+            name: "year, month, day",
+            reason: format!("invalid date: {}-{}-{}", year, month, day).into(),
+        });
+    }
+    let adjusted_year = if year < 0 && !for_timestamp {
+        // For a year `-X`, The `Date` type is printed as `X+1 BC`, while `Timestamp` is printed as `-X`.
+        year + 1
+    } else {
+        year
+    };
```
Comment (xiangjinwu) [inline src/expr/impl/src/scalar/make_time.rs:34]: Did not quite get this part. Date / timestamp / timestamptz shall all follow the same rule:
* In PG SQL, the year before 1 is -1. There is no year 0. It is printed as `1 BC`.
* In ISO 8601 (chrono crate), the year before 1 is 0. It can be printed as `1 BC` (`year_ce` instead of `year`)

Practically it is okay to be incompatible with PG for BC years. But the logic here is both wrong and more complicated

In short, `year + 1` as long as `year < 0` when passing a PG SQL input to the chrono crate. This yields the correct value in memory during input. Whether showing `X+1 BC` or `-X` shall be adjusted on the output side.

**Review by xiangjinwu — COMMENTED** [review]
Rest LGTM!

File: src/expr/impl/src/scalar/make_time.rs (hunk 1 above)
Comment (KeXiangWang) [reply src/expr/impl/src/scalar/make_time.rs]: After discussion with xiangjin, we conduct `+1` for both `Date` and `Timestamp`/`Timestamptz`. However, as our `ToText` for `Date` and `Timestamp`/`Timestamptz` are different, the output of `make_date` and `make_timestamp/timestamptz` are different.

```
SELECT make_timestamp(-5, 02, 29, 08, 15, 55.33);
----
-0004-02-29 08:15:55.330

SELECT make_date(-5, 2, 29);
----
0005-02-29 BC
```

We'll fix it later in another PR by modify the impl of `ToText` for`Timestamp`/`Timestamptz`.

**Review by xiangjinwu — APPROVED** [review]
LGTM

File: e2e_test/batch/basic/make_time.slt.part [hunk 2]
```diff
@@ -9,20 +9,31 @@ SELECT make_timestamptz(1973, 07, 15, 08, 15, 55.33);
 query T
 SELECT make_timestamptz(-1973, 07, 15, 08, 15, 55.33);
 ----
--1973-07-15 08:15:55.330+00:00
+-1972-07-15 08:15:55.330+00:00
 
-query error Invalid parameter sec: invalid sec
+query error Invalid parameter year, month, day: invalid date: -3-2-29
```
Comment (BugenZhao) [inline e2e_test/batch/basic/make_time.slt.part:14]: In postgres we have `date field value out of range: -4-02-29`. Note the `-4` here. Are we behaving correctly?

**Review by BugenZhao — COMMENTED** [review]
Rest LGTM

**Review by BugenZhao — APPROVED** [review]

File: e2e_test/batch/basic/make_time.slt.part (hunk 2 above)
Comment (KeXiangWang) [reply e2e_test/batch/basic/make_time.slt.part:14]:
```
> SELECT make_date(-4, 02, 29);
ERROR:  date field value out of range: -3-02-29
> SELECT make_timestamp(-4, 02, 29, 08, 15, 55.33);
ERROR:  date field value out of range: -4-02-29
```
Pg behaves differently on `make_date` and `make_timestamp`. Seems to be a misaligned design. So I think it's OK if we don't follow it strictly.
